File: Python/Python/Week_1/Day_2/Core/BankAccount/bankaccount.py
class BankAccount :
    all_accounts = []

def __init__(self, int_rate=0.01,balance=0):
    self.int_rate = int_rate
    self.balance = balance
    BankAccount.all_accounts.append(self)

def deposit(self,amount):
    self.balance += amount
    return self

def withdraw(self, amount):
    if self.balance >= amount:
        self.balance -= amount 
    else:
        print("Insufficient funds: Charging a $5 fee")
        self.balance -= 5
    return self 

def display_account_info(self):
    print(f"Balance: ${self.balance}")
    return self

def yield_interest(self):
    if self.balance > 0:
        self.balance += self.balance * self.int_rate
    return self 

@classmethod
def print_all_accounts(cls):
    print("All Accounts")
for account in cls.all_accounts:
    account.display_account_info()

compte1 = BankAccount(0.01, 100)
compte2 = BankAccount(0.02)
compte1.deposit(100).deposit(200).deposit(300).withdraw(50).yield_interest().display_account_info()

File: NewJava/Java_OOP/Week_1/Day_2/Core/BankAccount/BankAccount.java
import java.util.Random;

public class BankAccount {
    // MEMBER VARIABLES
    private double checkingBalance;
    private double savingsBalance;
    private String accountNumber;

    private static int accounts = 0;
    private static double totalMoney = 0.0;

    // CONSTRUCTOR
    public BankAccount() {
        this.checkingBalance = 0;
        this.savingsBalance = 0;
        this.accountNumber = generateAccountNumber();
        accounts++;
    }

    // PRIVATE METHOD (Ninja Bonus)
    private String generateAccountNumber() {
        Random rand = new Random();
        StringBuilder accNum = new StringBuilder();
        for (int i = 0; i < 10; i++) {
            accNum.append(rand.nextInt(10)); // generates digit 0-9
        }
        return accNum.toString();
    }

    // GETTERS
    public double getCheckingBalance() {
        return checkingBalance;
    }

    public double getSavingsBalance() {
        return savingsBalance;
    }

    public static int getAccounts() {
        return accounts;
    }

    public static double getTotalMoney() {
        return totalMoney;
    }

    public String getAccountNumber() {
        return accountNumber;
    }

    // DEPOSIT METHOD
    public void deposit(String accountType, double amount) {
        if (amount <= 0) {
            System.out.println("Deposit must be positive.");
            return;
        }

        if (accountType.equalsIgnoreCase("checking")) {
            checkingBalance += amount;
        } else if (accountType.equalsIgnoreCase("savings")) {
            savingsBalance += amount;
        } else {
            System.out.println("Invalid account type.");
            return;
        }
        totalMoney += amount;
    }

    // WITHDRAW METHOD
    public void withdraw(String accountType, double amount) {
        if (amount <= 0) {
            System.out.println("Withdrawal must be positive.");
            return;
        }

        if (accountType.equalsIgnoreCase("checking")) {
            if (checkingBalance >= amount) {
                checkingBalance -= amount;
                totalMoney -= amount;
            } else {
                System.out.println("Insufficient funds in checking.");
            }
        } else if (accountType.equalsIgnoreCase("savings")) {
            if (savingsBalance >= amount) {
                savingsBalance -= amount;
                totalMoney -= amount;
            } else {
                System.out.println("Insufficient funds in savings.");
            }
        } else {
            System.out.println("Invalid account type.");
        }
    }

    // BALANCE METHOD
    public void displayBalance() {
        System.out.printf("Account %s | Checking: $%.2f | Savings: $%.2f | Total: $%.2f\n", 
                        accountNumber, checkingBalance, savingsBalance, checkingBalance + savingsBalance);
    }
}
